lib/index.js

```diff
@@ -70,33 +70,47 @@ Intercom.prototype.identify = function(identify, fn){
 };
 
 /**
- * Group in two steps - company and then user
+ * Group call: upsert company data for users via Intercom Bulk User API
+ * Even though this is not explicitly an endpoint for their `/companies` resource,
+ * we can send `companies` object via Bulk User API (recommended by Intercom directly)
+ *
+ * https://developers.intercom.com/reference#bulk-user-operations
+ *
+ * To preserve as much order as possible for multiple `.group()` calls for a given userId we will:
+ *
+ * 1) Lookup in redis if there is an existing job for a `userId`
+ * 2) If exists, add to the job
+ * 3) Otherwise create a new job, save jobId in redis with expiration
+ * 4) Error handle in case we try to add to an invalid job
  *
  * @param {Group} group
  * @param {Function} fn
  * @api public
  */
 
 Intercom.prototype.group = function(group, fn){
-  var json = group.json();
-  var traits = json.traits || {};
+  var traits = group.traits();
+  var id = group.userId() || group.email();
+  // We are locking separately from `.identify()` calls by adding 'groups'
+  // Since both endpoints are upsertions so race condition does not matter.
+  // If there is no profile yet for this message's userId, it will create it under
+  // the ensured `userId` field, and a later `.identify()` call with the user's traits will be updated
+  var key = [this.settings.appId, 'groups', id].join(':');
   var self = this;
 
-  return this
-    .post('/companies')
-    .set(headers(group, this.settings))
-    .type('json')
-    .accept('json')
-    .send(mapper.group(group))
-    .end(this.handle(function(err){
-        if (err) return fn(err);
-        json.userId = group.userId();
-        traits.id = group.groupId();
-        json.traits = { companies: [traits] };
-        var identify = new Identify(json);
-        self.identify(identify, fn);
-      })
-    );
+  this.lock(key, function(err){
+    if (err) return fn(err);
+    var redis = self.redis();
+    var jobKey = [self.settings.appId, 'jobs', id].join(':');
```

Comment:
This needs to be a different job than the `track` calls right? We need a different key here then